app/scripts/controllers/permissions/background-api.test.js

```diff
@@ -1,197 +1,620 @@
+                'eip155:1:0x3',
+                'eip155:1:0x1',
+                'eip155:1:0x4',
+              ],
+            },
+          },
+          isMultichainOrigin: true,
+        },
+      );
     });
   });
 
   describe('addMorePermittedAccounts', () => {
-    it('calls grantPermissionsIncremental with expected parameters for single account', () => {
+    it('gets the CAIP-25 caveat', () => {
       const permissionController = {
-        grantPermissionsIncremental: jest.fn(),
+        getCaveat: jest.fn(),
       };
 
-      getPermissionBackgroundApiMethods(
-        permissionController,
-      ).addMorePermittedAccounts('foo.com', ['0x1']);
+      try {
+        getPermissionBackgroundApiMethods({
+          permissionController,
+        }).addMorePermittedAccounts('foo.com', ['0x1']);
+      } catch (err) {
+        // noop
+      }
 
-      expect(
-        permissionController.grantPermissionsIncremental,
-      ).toHaveBeenCalledTimes(1);
-      expect(
-        permissionController.grantPermissionsIncremental,
-      ).toHaveBeenCalledWith({
-        subject: { origin: 'foo.com' },
-        approvedPermissions: getApprovedPermissions(['0x1']),
-      });
+      expect(permissionController.getCaveat).toHaveBeenCalledWith(
+        'foo.com',
+        Caip25EndowmentPermissionName,
+        Caip25CaveatType,
+      );
+    });
+
+    it('throws an error if there is no existing CAIP-25 caveat', () => {
+      const permissionController = {
+        getCaveat: jest.fn(),
+      };
+
+      expect(() =>
+        getPermissionBackgroundApiMethods({
+          permissionController,
+        }).addMorePermittedAccounts('foo.com', ['0x1']),
+      ).toThrow(
+        new Error('tried to add accounts when none have been permissioned'),
+      );
     });
 
-    it('calls grantPermissionsIncremental with expected parameters with multiple accounts', () => {
+    it('calls updateCaveat with the accounts added and all eip155 accounts synced', () => {
       const permissionController = {
-        grantPermissionsIncremental: jest.fn(),
+        getCaveat: jest.fn().mockReturnValue({
+          value: {
+            requiredScopes: {
+              'eip155:1': {
+                methods: [],
+                notifications: [],
+              },
+              'eip155:10': {
+                methods: [],
+                notifications: [],
+                accounts: ['eip155:10:0x1', 'eip155:10:0x2'],
+              },
+            },
+            optionalScopes: {
+              'bip122:000000000019d6689c085ae165831e93': {
+                methods: [],
+                notifications: [],
+                accounts: [
+                  'bip122:000000000019d6689c085ae165831e93:128Lkh3S7CkDTBZ8W7BbpsN3YYizJMp8p6',
+                ],
+              },
+              'eip155:1': {
+                methods: [],
+                notifications: [],
+                accounts: ['eip155:1:0x2', 'eip155:1:0x3'],
+              },
+            },
```

Comment:
So part of whats happening here (perhaps what you're calling out with the "syncing") is, if somehow some accounts are not synced across all scopes - shouldn't be possible? - calling this `addMorePermittedAccounts` will sync them up?

Reply:
nit -> maybe worth calling out that it will not modify non EIP155 scopes

Reply:
> if somehow some accounts are not synced across all scopes - shouldn't be possible? 

In this PR it is possible for a scope to be added without the accounts in the permitted chain flow. See https://github.com/MetaMask/metamask-extension/pull/26483#discussion_r1739493377